### PR TITLE
fix(i18n): complete missing translations

### LINKED ISSUE #88: Localize labels (I18N)
Support localization for label. We should closely match the [bpmn-js implementation](https://github.com/bpmn-io/bpmn-js/issues/491).

__Tasks__

* [ ] Labels are piped through a `translate` helper to provide a simple localization stub

### PR DESCRIPTION
**Pull request by PTYin**
Hi there👋, I'm currently working on a project that utilizes dmn-js to build an online DMN modeler. The code of dmn-js is organized in a clear and concise manner. Thanks for your fantastic works. 

However, the project that I'm working on is oriented towards Chinese users, and I found the full coverage of i18n for dmn-js is still not implemented. Thus I have written this pr to address the problem (Closes #88).

The newly added translations are listed at the bottom of this pr.

Except for code modification to add translations, I have also modified `packages/dmn-js/test/spec/TabbingSpec.js` to enable check for missing translations. You can use the following command to check:

```shell
TEST_BROWSERS=Chrome;SINGLE_START=translation npm run dev -- dmn-js
```

Every text is shown as `tr(xxx)` if it's translatable:

image

# ADDED TRANSLATIONS

## DRD

- Definition Properties
  - Title for Decision Name
  - Title for Decision ID
  - DecisionIdEdit Error Message

## Decision Table

- Decision Table Properties
  - Title for Decision Name
- Hit Policy
  - Label
  - Value
  - Explanation
  - Options
- Table Head
  - Input Cell
    - Type Ref
  - Input Cell Context Menu
    - Placeholder for Enter Expression
  - Output Cell
    - Type Ref
  - Allowed Values Editing
    - Placeholder for Predefined Values
    - Label for Nonempty Predefined Values
  - Type Ref Cell Context Menu
    - Options for Type Ref
- Simple Mode Button
  - Title
- Simple Boolean Edit
  - Heading
  - Options
  - Placeholder
  - Error Message
- Simple Date Edit
  - Heading
  - Options
  - Validation
- Simple Date Time Edit
  - Heading
  - Options
  - Validation
- Simple Duration Edit
  - Heading
  - Options
  - Validation
- Simple Number Edit
  - Heading
  - Options
  - Validation
- Simple String Edit
  - Input Options (match one / match none)
  - Label for Predefined Values
  - Label for Custom Values
  - Placeholder for Adding Values or Setting Value
- Simple Time Edit
  - Heading
  - Options
  - Validation
- View Drd Button
  - Button Text

## Literal Expression

- Literal Expression Properties
  - Label for Variable Name
  - Label for Variable Type
  - Options for Variable Type Ref
  - Label for Expression Language
- View Drd Button
  - Button Text

### CODE REVIEW

File: packages/dmn-js-decision-table/src/features/simple-string-edit/components/SimpleStringEditContextMenuComponent.js [hunk 1]
```diff
@@ -254,7 +256,7 @@ export default class SimpleStringEditContextMenuComponent extends Component {
         value,
         isChecked: true,
         isRemovable: true,
-        group: 'Custom Values'
+        group: this._translate(INPUT_ENTRY_VALUES_LABEL)
```
Comment (nikku) [inline packages/dmn-js-decision-table/src/features/simple-string-edit/components/SimpleStringEditContextMenuComponent.js:259]:
```suggestion
        group: this._translate('Custom Values')

```

File: packages/dmn-js-decision-table/src/features/simple-time-edit/components/InputTimeEdit.js [hunk 2]
```diff
@@ -208,15 +213,18 @@ export default class InputTimeEdit extends Component {
               <ValidatedInput
                 className="end-time-input dms-block"
                 onInput={ this.onEndTimeInput }
-                placeholder={ `e.g. ${ getSampleTime() }` }
-                validate={ validateISOString }
+                placeholder={ `${this._translate('e.g.')} ${ getSampleDate() }` }
```
Comment (nikku) [inline packages/dmn-js-decision-table/src/features/simple-time-edit/components/InputTimeEdit.js:216]:
```suggestion
                placeholder={ this._translate('e.g. { example } ', { example: getSampleDate() }) }
```

**Review by nikku — CHANGES_REQUESTED** [review]
Incredible contribution, thank you!

I just found two small changes that you may want to incorporate.

Will give it a detailed test run in just a bit.

**Comment by nikku** [comment]
Let's add the test line you mentioned as a new script to the `package.json`:

```
"scripts": {
  ...
  "start:translated": "cross-env SINGLE_START=translation npm run dev -- dmn-js",
  ...
}
```

File: packages/dmn-js-decision-table/src/features/decision-table-properties/components/DecisionTablePropertiesComponent.js [hunk 3]
```diff
@@ -25,7 +27,8 @@ export default class DecisionTablePropertiesComponent extends Component {
 
     return (
       <div className="decision-table-properties">
-        <div className="decision-table-name" title={ 'Decision Name: ' + name }>
```
Comment (nikku) [inline packages/dmn-js-decision-table/src/features/decision-table-properties/components/DecisionTablePropertiesComponent.js:28]: Is there a good reason (translation wise) why we translate the `:` with `Decision Name`?

File: packages/dmn-js-decision-table/src/features/simple-date-edit/components/InputDateEdit.js [hunk 4]
```diff
@@ -175,30 +176,33 @@ export default class InputDateEdit extends Component {
...
           }
         </h4>
 
         <div>
           <ValidatedInput
             className="start-date-input dms-block"
             onInput={ this.onStartDateInput }
-            placeholder={ `e.g. ${ getSampleDate() }` }
-            validate={ validateISOString }
+            placeholder={ `${this._translate('e.g.')} ${ getSampleDate() }` }
```
Comment (nikku) [inline packages/dmn-js-decision-table/src/features/simple-date-edit/components/InputDateEdit.js:188]:
```suggestion
            placeholder={ this._translate('e.g. { sample }', { sample: getSampleDate() }) }

```

File: packages/dmn-js-decision-table/src/features/simple-date-edit/components/InputDateEdit.js [hunk 5]
```diff
@@ -208,15 +212,18 @@ export default class InputDateEdit extends Component {
               <ValidatedInput
                 className="end-date-input dms-block"
                 onInput={ this.onEndDateInput }
-                placeholder={ `e.g. ${ getSampleDate() }` }
-                validate={ validateISOString }
+                placeholder={ `${this._translate('e.g.')} ${ getSampleDate() }` }
```
Comment (nikku) [inline packages/dmn-js-decision-table/src/features/simple-date-edit/components/InputDateEdit.js:215]:
```suggestion
                placeholder={ this._translate('e.g. { sample }', { sample: getSampleDate() }) }

```

File: packages/dmn-js-decision-table/src/features/simple-date-edit/components/OutputDateEdit.js [hunk 6]
```diff
@@ -65,23 +66,24 @@ export default class OutputDateEdit extends Component {
...
 
-        <h4 class="dms-heading">Set date</h4>
+        <h4 class="dms-heading">{ this._translate('Set date') }</h4>
 
         <div>
           <ValidatedInput
             onInput={ this.onInput }
-            placeholder={ `e.g. ${ getSampleDate() }` }
-            validate={ validateISOString }
+            placeholder={ `${ this._translate('e.g.') } ${ getSampleDate() }` }
```
Comment (nikku) [inline packages/dmn-js-decision-table/src/features/simple-date-edit/components/OutputDateEdit.js:76]:
```suggestion
            placeholder={ this._translate('e.g. { sample }', { sample: getSampleDate() }) }

```

File: packages/dmn-js-decision-table/src/features/simple-date-time-edit/components/InputDateTimeEdit.js [hunk 7]
```diff
@@ -175,30 +176,33 @@ export default class InputDateEdit extends Component {
...
           }
         </h4>
 
         <div>
           <ValidatedInput
             className="start-date-input dms-block"
             onInput={ this.onStartDateInput }
-            placeholder={ `e.g. ${ getSampleDate() }` }
-            validate={ validateISOString }
+            placeholder={ `${ this._translate('e.g.') } ${ getSampleDate() }` }
```
Comment (nikku) [inline packages/dmn-js-decision-table/src/features/simple-date-time-edit/components/InputDateTimeEdit.js:188]:
```suggestion
            placeholder={ this._translate('e.g. { sample }', { sample: getSampleDate() }) }
```

File: packages/dmn-js-decision-table/src/features/simple-date-time-edit/components/InputDateTimeEdit.js [hunk 8]
```diff
@@ -208,15 +212,18 @@ export default class InputDateEdit extends Component {
               <ValidatedInput
                 className="end-date-input dms-block"
                 onInput={ this.onEndDateInput }
-                placeholder={ `e.g. ${ getSampleDate() }` }
-                validate={ validateISOString }
+                placeholder={ `${ this._translate('e.g.') } ${ getSampleDate() }` }
```
Comment (nikku) [inline packages/dmn-js-decision-table/src/features/simple-date-time-edit/components/InputDateTimeEdit.js:215]:
```suggestion
                placeholder={ this._translate('e.g. { sample }', { sample: getSampleDate() }) }
```

File: packages/dmn-js-decision-table/src/features/simple-date-time-edit/components/OutputDateTimeEdit.js [hunk 9]
```diff
@@ -65,23 +66,24 @@ export default class OutputDateEdit extends Component {
...
 
-        <h4 class="dms-heading">Set date and time</h4>
+        <h4 class="dms-heading">{ this._translate('Set date and time') }</h4>
 
         <div>
           <ValidatedInput
             onInput={ this.onInput }
-            placeholder={ `e.g. ${ getSampleDate() }` }
-            validate={ validateISOString }
+            placeholder={ `${ this._translate('e.g.') } ${ getSampleDate() }` }
```
Comment (nikku) [inline packages/dmn-js-decision-table/src/features/simple-date-time-edit/components/OutputDateTimeEdit.js:76]:
```suggestion
            placeholder={ `${ this._translate('e.g. { sample }', { sample: getSampleDate() }) }
```

File: packages/dmn-js-decision-table/src/features/simple-duration-edit/components/DurationInput.js [hunk 10]
```diff
@@ -34,15 +35,15 @@ export class DurationInput extends Component {
...
     if (!validateDuration(this._type, value)) {
-      return ERROR_MESSAGE[this._type];
+      return this._translate(ERROR_MESSAGE[this._type]);
     }
   }
 
   _getPlaceholder() {
     if (this._type === 'yearMonthDuration') {
-      return 'e.g. P1Y2M';
+      return this._translate('e.g.') + ' P1Y2M';
```
Comment (nikku) [inline packages/dmn-js-decision-table/src/features/simple-duration-edit/components/DurationInput.js:44]:
```suggestion
      return this._translate('e.g. { sample }', { sample: 'P1Y2M' });
```

File: packages/dmn-js-decision-table/src/features/simple-duration-edit/components/DurationInput.js [hunk 11]
```diff
@@ -34,15 +35,15 @@ export class DurationInput extends Component {
...
     } else if (this._type === 'dayTimeDuration') {
-      return 'e.g. P1DT2H';
+      return this._translate('e.g.') + ' P1DT2H';
```
Comment (nikku) [inline packages/dmn-js-decision-table/src/features/simple-duration-edit/components/DurationInput.js:46]:
```suggestion
      return this._translate('e.g. { sample }', { sample: 'P1DT2H' });
```

File: packages/dmn-js-decision-table/src/features/simple-time-edit/components/InputTimeEdit.js [hunk 12]
```diff
@@ -175,30 +177,33 @@ export default class InputTimeEdit extends Component {
...
           }
         </h4>
 
         <div>
           <ValidatedInput
             className="start-time-input dms-block"
             onInput={ this.onStartTimeInput }
-            placeholder={ `e.g. ${ getSampleTime() }` }
-            validate={ validateISOString }
+            placeholder={ `${this._translate('e.g.')} ${ getSampleDate() }` }
```
Comment (nikku) [inline packages/dmn-js-decision-table/src/features/simple-time-edit/components/InputTimeEdit.js:189]:
```suggestion
            placeholder={ this._translate('e.g. { sample }', { sample: getSampleDate() }) }
```

File: packages/dmn-js-decision-table/src/features/simple-time-edit/components/OutputTimeEdit.js [hunk 13]
```diff
@@ -65,23 +67,24 @@ export default class OutputTimeEdit extends Component {
...
 
-        <h4 class="dms-heading">Set Date</h4>
+        <h4 class="dms-heading">{ this._translate('Set Date') }</h4>
 
         <div>
           <ValidatedInput
             onInput={ this.onInput }
-            placeholder={ `e.g. ${ getSampleTime() }` }
-            validate={ validateISOString }
+            placeholder={ `${this._translate('e.g.')} ${ getSampleDate() }` }
```
Comment (nikku) [inline packages/dmn-js-decision-table/src/features/simple-time-edit/components/OutputTimeEdit.js:77]:
```suggestion
            placeholder={ this._translate('e.g. { sample }', { sample: getSampleDate() }) }
```

File: packages/dmn-js-decision-table/src/features/decision-table-properties/components/DecisionTablePropertiesComponent.js (hunk 3 above)
Comment (PTYin) [reply packages/dmn-js-decision-table/src/features/decision-table-properties/components/DecisionTablePropertiesComponent.js:28]: There exists a need to replace the colon `:` with Chinese colon `：`(\uff1a). Besides, I suppose punctuation is commonly included according to this stackexchange answer ([link](https://softwareengineering.stackexchange.com/a/374009))

**Comment by PTYin** [comment]
THX, I have modified the commit according to suggestions except for the colon one. About the colon inclusion problem, please read my comment above.

**Review by nikku — COMMENTED** [review]
Good stuff, works like a charm.

**Comment by nikku** [comment]
Released with v12.3.0. Thanks again :clap:

**Pull request merged**